File: Leetcode/sort_and_search/33_Search in Rotated Sorted Array/Solution.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums.length==0)
            return -1;
        int low = 0, high = nums.length-1;
        while (low <= high){
            int mid = (low+high)/2;
            if (nums[mid] == target){
                return mid;
            }
            if (nums[mid]>=nums[low]){
                if (target<nums[mid] && target>=nums[low])
                    high = mid-1;
                else
                    low = mid+1;
            }
            if (nums[mid]<=nums[high]){
                if (target>nums[mid] && target<=nums[high])
                    low = mid+1;
                else
                    high = mid-1;
            }
        }
        return -1;
    }
}

File: Leetcode/sort_and_search/33_Search in Rotated Sorted Array/Solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        return self.rotated_bisect(0, len(nums)-1, nums, target)

    def rotated_bisect(self, begin, end, nums,target):
        if begin>end:
            return -1
        mid = begin + (end-begin)//2
        if nums[mid]==target:
            return mid
        if nums[begin] <= nums[mid]:        # the first half is sorted
            if nums[begin]<=target<=nums[mid]:
                return self.rotated_bisect(begin, mid-1, nums, target)
            return self.rotated_bisect(mid+1, end, nums, target)
        else:
            if nums[mid]<=target<=nums[end]:
                return self.rotated_bisect(mid+1, end, nums, target)
            else:
                return self.rotated_bisect(begin, mid-1, nums, target)
